crates/wit-component/src/lib.rs

```diff
@@ -79,6 +79,43 @@ impl From<StringEncoding> for wasm_encoder::CanonicalOption {
     }
 }
 
+/// Handles world name resolution for cases when multiple packages may have been resolved. If this
+/// is the case, and we're dealing with input that contains a user-supplied world name (like via a
+/// CLI command, for instance), we want to ensure that the world name follows the following rules:
+///
+///   * If there is a single resolved package with a single world, the world name name MAY be
+///     omitted.
+///   * If there is a single resolved package with multiple worlds, the world name MUST be supplied,
+///     but MAY or MAY NOT be fully-qualified.
+///   * If there are multiple resolved packages, the world name MUST be fully-qualified.
+pub fn resolve_world_from_name(
```

Comment:
This is more-or-less a replacement of `Resolve::select_world`. In that case what would you think of basically replacing `select_world` with this? That way callers can largely stay the same where they'll pass in a list of ids here rather than a single id and the `"foo"` selector will only work if the list of packages is of length 1.

Reply:
I initially had the same instinct, but felt like the current approach may be a little more typesafe?

In either case, the caller has to make sure of something at the call site: in the current setup, they have to make sure to select the right `PackageId` for the `world_name` they're interested in (I can add a comment to the doc-string to help point them to the right helper for doing this), but at least the type signature gives a hint that they need to do this. In the alternative world where `select_world` always takes a `Vec<PackageId>`, I could easily imagine authors forgetting to check that their input `world_name` formatting conforms to the rules dictated by the `Vec<PackageId>` count, and hitting this error condition in (probably relatively rarely tested) multi package scenarios.

Reply:
I think you're right, but I also think we'll still want to change `select_world`, lemme expand on this though. You're right in that if a single `PackageId` was what the caller already had then it's best to have the current signature of `select_world`. Note though that the `world` argument of `select_world` is not just `"foo"` but it can also be the fully qualified name of `foo:bar/baz`.

Otherwise though, why I think this still needs to change, is that the `select_world` is a method that has not aged well. The intention of the method was to be "here are the two inputs bindings generators typically have and then here's the canonical way to acquire a world". In that sense this is supposed to be **the** helper to acquire a `WorldId` to generate bindings for (and all bindings generators operate at the level of a world). The history of this method looked like:

* Originally there weren't worlds
* Then there were worlds, but there was a `default world` in each package
* Then we wanted the option to select a non-default world, hence the `world: Option<&str>` argument
* Then we removed `default world` but the `Option` remained for "if there's only one world otherwise disambiguate with the name here"
* Then we wanted to select worlds outside of this package so the string syntax was extended to `foo:bar/baz`.

Basically what I'm getting at is that the exact signature of `select_world` is not really all that important because it's pretty rare to have a single `PackageId` after this change. The input from all bindings generators is going to be "here's the set of packages that was defined in the 'main package' now please select a world from here". Given that I think it's best to switch this to `&[PackageId]`. I'd love one day to be able to remove the package argument entirely, but I think that's for a bit further down the road.

Reply:
If possible, I'd like to save such a refactor for a follow on PR, to get a good baseline landed and free of merge conflicts here.

Reply:
Sounds good to me 👍